v2/tools/generator/internal/astmodel/type_walker.go

```diff
@@ -45,41 +47,43 @@ type typeWalkerState struct {
 // NewTypeWalker returns a TypeWalker.
 // The provided visitor visitTypeName function must return a TypeName and visitObjectType must return an ObjectType or calls to
 // Walk will panic.
-func NewTypeWalker(allDefs TypeDefinitionSet, visitor TypeVisitor) *TypeWalker {
-	typeWalker := TypeWalker{
-		allDefinitions:          allDefs,
-		originalVisitTypeName:   visitor.visitTypeName,
-		originalVisitObjectType: visitor.visitObjectType,
-		AfterVisit:              IdentityAfterVisit,
-		MakeContext:             IdentityMakeContext,
-		ShouldRemoveCycle:       IdentityShouldRemoveCycle,
+func NewTypeWalker[C any](allDefs TypeDefinitionSet, visitor TypeVisitor[C]) *TypeWalker[C] {
+	typeWalker := TypeWalker[C]{
+		allDefinitions:                allDefs,
+		originalVisitInternalTypeName: visitor.visitInternalTypeName,
+		originalVisitObjectType:       visitor.visitObjectType,
+		AfterVisit:                    IdentityAfterVisit[C],
+		MakeContext:                   IdentityMakeContext[C],
+		ShouldRemoveCycle:             IdentityShouldRemoveCycle[C],
 	}
 
 	// visitor is a copy - modifications won't impact passed visitor
-	visitor.visitTypeName = typeWalker.visitTypeName
-	visitor.visitTypeNameIsIdentity = false // disable short-cut optimization
-
+	visitor.visitInternalTypeName = typeWalker.visitInternalTypeName
 	visitor.visitObjectType = typeWalker.visitObjectType
 
 	typeWalker.visitor = visitor
 
 	return &typeWalker
 }
 
-func (t *TypeWalker) visitTypeName(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error) {
+func (t *TypeWalker[C]) visitInternalTypeName(this *TypeVisitor[C], it InternalTypeName, ctx C) (Type, error) {
 	updatedCtx, err := t.MakeContext(it, ctx)
 	if err != nil {
 		return nil, errors.Wrapf(err, "MakeContext failed for name %q", it)
 	}
 
-	visitedTypeName, err := t.originalVisitTypeName(this, it, updatedCtx)
-	if err != nil {
-		return nil, errors.Wrapf(err, "visitTypeName failed for name %q", it)
-	}
+	if t.originalVisitInternalTypeName != nil {
+		visitedTypeName, terr := t.originalVisitInternalTypeName(this, it, updatedCtx)
+		if terr != nil {
+			return nil, errors.Wrapf(terr, "visitTypeName failed for name %q", it)
+		}
 
-	it, ok := visitedTypeName.(TypeName)
-	if !ok {
-		panic(fmt.Sprintf("TypeWalker visitor visitTypeName must return a TypeName, instead returned %T", visitedTypeName))
+		itn, ok := visitedTypeName.(InternalTypeName)
+		if !ok {
+			panic(fmt.Sprintf("TypeWalker visitor visitTypeName must return a TypeName, instead returned %T", visitedTypeName))
```

Comment:
minor: Fix comment here? We're checking explicitly for `InternalTypeName`, so should say that?

Reply:
Done.